worker/rhpv2.go

```diff
@@ -717,6 +734,60 @@ func (w *worker) withRevisionV2(lockTimeout time.Duration, t *rhpv2.Transport, h
 	return fn(t, rev, settings)
 }
 
+// compatPruneHostV021RetryCheck is an error check that is used to determine
+// whether we want to retry a prune operation on a host that is running an older
+// version of the host software that calculates costs differently. This method
+// returns true if and only if the error indicates an insufficient host
+// transfer, but also the 'got' matches what we sent and the 'want' matches what
+// we would send in a retry
+func (w *worker) compatPruneHostV021RetryCheck(err error, hk types.PublicKey, hs rhpv2.HostSettings, sent types.Currency, actions uint64) bool {
+	w.pruneMu.Lock()
+	defer w.pruneMu.Unlock()
+
+	// check if we're dealing with insufficient host transfer
+	if !utils.IsErr(err, errInsufficientHostTransfer) {
+		return false
+	}
+
+	// check if we're dealing with a known host, this is important because it
+	// ensures we don't keep retrying should we receive the same error on retry
+	if _, found := w.pruneCompatHostV021[hk]; found {
+		return false
+	}
+
+	// check if we can parse two currency values from the error message (got/want)
+	matches := regexp.MustCompile("\\d+").FindAllString(err.Error(), -1)
+	if len(matches) != 2 {
+		return false
+	}
+
+	// check if 'got' matches what we went
+	got, err := types.ParseCurrency(matches[1])
+	if err != nil {
+		return false
+	} else if !got.Equals(sent) {
+		return false
+	}
+
+	// check if 'want' matches what we would send in a retry
+	want, err := types.ParseCurrency(matches[0])
+	if err != nil {
+		return false
+	} else if want.Cmp(hs.BaseRPCPrice.Add(hs.DownloadBandwidthPrice.Mul64(crypto.HashSize*(128+actions)))) != 0 {
```

Comment:
How big is the difference in prices? Because if it is negligible, a much easier solution would be to just compute both and pick the larger value until we have rhpv4 and drop rhpv2.

Reply:
Yeah it's a lot more expensive, that's all relative of course. In the description I mentioned the 40TiB contracts so we're paying for 40TiB proofs essentially. The cost is roughly 10x... so I figured this retry was worth it.